File: apps/web/lib/training.ts

// apps/web/lib/training.ts

import { config } from "@/lib/config";
import { getAuthHeaders } from "./auth-headers";

export type TrainingExerciseInput = {
    exerciseId: string;
    exerciseName: string;
    sets: number;
    reps?: number;
    weightKg?: number;
    notes?: string;
};

export type RunningInput = {
    distanceKm: number;
    timeMinutes: number;
};

export type TrainingLog = {
    id: string;
    date: string;
    exercises: TrainingExerciseInput[];
    running?: RunningInput | null;
    createdAt: string;
    updatedAt: string;
};

export type TrainingLogUpsertRequest = {
    exercises: TrainingExerciseInput[];
    running?: RunningInput | null;
};

export async function getTrainingLog(date: string): Promise<TrainingLog | null> {
    const res = await fetch(`${config.apiBaseUrl}/training/logs/${date}`, {
        headers: await getAuthHeaders(),
    });

    if (!res.ok) {
        throw new Error("Failed to load training log");
    }

    return res.json();
}

export async function saveTrainingLog(
    date: string,
    body: TrainingLogUpsertRequest
): Promise<TrainingLog> {
    const res = await fetch(`${config.apiBaseUrl}/training/logs/${date}`, {
        method: "PUT",
        headers: {
            "Content-Type": "application/json",
            ...(await getAuthHeaders()),
        },
        body: JSON.stringify(body),
    });

    if (!res.ok) {
        throw new Error("Failed to save training log");
    }

    return res.json();
}

export type TrainingTemplate = {
    id: string;
    name: string;
    exercises: TrainingExerciseInput[];
    createdAt: string;
    updatedAt: string;
};

export type TrainingTemplateCreateRequest = {
    name: string;
    exercises: TrainingExerciseInput[];
};

export async function getTrainingTemplates(): Promise<TrainingTemplate[]> {
    const res = await fetch(`${config.apiBaseUrl}/training/templates`, {
        headers: await getAuthHeaders(),
    });

    if (!res.ok) {
        throw new Error("Failed to load training templates");
    }

    return res.json();
}

export async function createTrainingTemplate(
    body: TrainingTemplateCreateRequest
): Promise<TrainingTemplate> {
    const res = await fetch(`${config.apiBaseUrl}/training/templates`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            ...(await getAuthHeaders()),
        },
        body: JSON.stringify(body),
    });

    if (!res.ok) {
        throw new Error("Failed to create training template");
    }

    return res.json();
}

export async function deleteTrainingTemplate(id: string): Promise<void> {
    const res = await fetch(`${config.apiBaseUrl}/training/templates/${id}`, {
        method: "DELETE",
        headers: await getAuthHeaders(),
    });

    if (!res.ok) {
        throw new Error("Failed to delete training template");
    }
}

export async function getTrainingLogsForMonth(
    month: string
): Promise<TrainingLog[]> {
    const res = await fetch(`${config.apiBaseUrl}/training/logs?month=${month}`, {
        headers: await getAuthHeaders(),
    });

    if (!res.ok) {
        throw new Error("Failed to load monthly training logs");
    }

    return res.json();
}

File: apps/api/app/schemas/training.py

# apps/api/app/schemas/training.py

from datetime import date, datetime
from pydantic import BaseModel, Field


class TrainingExerciseInput(BaseModel):
    exerciseId: str
    exerciseName: str
    sets: int = Field(ge=1)
    reps: int | None = Field(default=None, ge=1)
    weightKg: float | None = Field(default=None, ge=0)
    notes: str | None = None


class RunningInput(BaseModel):
    distanceKm: float = Field(ge=0)
    timeMinutes: float = Field(ge=0)

class DistanceActivityInput(BaseModel):
    distanceKm: float = Field(ge=0)
    timeMinutes: float = Field(ge=0)


class MobilityInput(BaseModel):
    timeMinutes: float = Field(ge=0)
    notes: str | None = None


class TrainingLogUpsertRequest(BaseModel):
    exercises: list[TrainingExerciseInput] = []
    running: DistanceActivityInput | None = None
    walking: DistanceActivityInput | None = None
    swimming: DistanceActivityInput | None = None
    cycling: DistanceActivityInput | None = None
    mobility: MobilityInput | None = None


class TrainingLogResponse(BaseModel):
    id: str
    date: date
    exercises: list[TrainingExerciseInput]
    running: DistanceActivityInput | None = None
    walking: DistanceActivityInput | None = None
    swimming: DistanceActivityInput | None = None
    cycling: DistanceActivityInput | None = None
    mobility: MobilityInput | None = None

    createdAt: datetime
    updatedAt: datetime

    model_config = {
        "from_attributes": True
    }


class TrainingTemplateCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=80)
    exercises: list[TrainingExerciseInput] = []

    running: DistanceActivityInput | None = None
    walking: DistanceActivityInput | None = None
    swimming: DistanceActivityInput | None = None
    cycling: DistanceActivityInput | None = None
    mobility: MobilityInput | None = None


class TrainingTemplateResponse(BaseModel):
    id: str
    name: str
    exercises: list[TrainingExerciseInput]
    running: DistanceActivityInput | None = None
    walking: DistanceActivityInput | None = None
    swimming: DistanceActivityInput | None = None
    cycling: DistanceActivityInput | None = None
    mobility: MobilityInput | None = None

    createdAt: datetime
    updatedAt: datetime